Issue #15685: fixed JavadocParagraph to detect paragraph tag when they have their corresponding closing tag

Zopsss opened the pull request:
fixes #15685 https://github.com/checkstyle/checkstyle/issues/15732

tried fixing the implementation of JavadocParagraph ~might have introduced new false-negatives :p~

I tried to make the Check work same way when the paragraph tag's corresponding closing tag is also present.

when I tried to run build command locally, I got so many errors for different files like checks, filters, etc for `<p>` tag. Most of them were for:

```
<p> tag should be placed immediately before the first word, with no space after.
```

@romani need your help here

---

Diff Regression config: https://gist.githubusercontent.com/Zopsss/013f379d4c821f2d624be56212a64575/raw/5f17ec9186d4421396429ab38cbac3e4663bf4eb/javadocparagraph_config.xml
Diff Regression projects: https://raw.githubusercontent.com/checkstyle/contribution/master/checkstyle-tester/projects-to-test-on-for-github-action.properties

romani commented:
we need to use for checkstyle-checks.xml 
`<property name="allowNewlineParagraph" value="false"/>`

Zopsss commented:
I have set the `allowNewlineParagraph` to false in checkstyle-checks.xml and now there is no violation for

```
<p> tag should be placed immediately before the first word, with no space after.
```

though there are still following errors in the build failure:

```
<p> tag should be preceded with an empty line.
Redundant <p> tag.
```

Locally I tried to fix `allowNewlineParagraph` by doing something like this:

```java

    /**
     * Tests whether the paragraph tag is misplaced.
     * Meaning that it is not immediate before the first word or another tag.
     *
     * @param tag html tag.
     * @return true, if the paragraph tag is misplaced.
     */
    private boolean isMisplacedParagraph(DetailNode tag) {
        boolean shouldGiveViolation;

        if (allowNewlineParagraph) {
            shouldGiveViolation = false;
        }
        else if (JavadocUtil.getFirstChild(tag).getType() == JavadocTokenTypes.PARAGRAPH) {
            final DetailNode paragraphToken = JavadocUtil.getFirstChild(tag);
            final DetailNode paragraphStartTagToken = JavadocUtil.getFirstChild(paragraphToken);
            final DetailNode nextSibling = JavadocUtil.getNextSibling(paragraphStartTagToken);
            shouldGiveViolation = nextSibling.getType() == JavadocTokenTypes.NEWLINE
                    || nextSibling.getType() == JavadocTokenTypes.EOF
                    || nextSibling.getText().startsWith(" ");
        }
        else {
            final DetailNode nextSibling = JavadocUtil.getNextSibling(tag);
            shouldGiveViolation = nextSibling.getType() == JavadocTokenTypes.NEWLINE
                    || nextSibling.getType() == JavadocTokenTypes.EOF
                    || nextSibling.getText().startsWith(" ");
        }

        return shouldGiveViolation;
    }
```

```java

    /**
     * Determines whether or not the line with paragraph tag has previous empty line.
     *
     * @param tag html tag.
     */
    private void checkParagraphTag(DetailNode tag) {
        final DetailNode newLine = getNearestEmptyLine(tag);
        if (isFirstParagraph(tag)) {
            log(tag.getLineNumber(), MSG_REDUNDANT_PARAGRAPH);
        }
        else if (newLine == null || tag.getLineNumber() - newLine.getLineNumber() != 1) {
            log(tag.getLineNumber(), MSG_LINE_BEFORE);
        }
        if (isMisplacedParagraph(tag)) { // this line is changed.
            log(tag.getLineNumber(), MSG_MISPLACED_TAG);
        }
    }
```

With this there was no error of:

```
<p> tag should be placed immediately before the first word, with no space after.
```

and I also didn't needed to set `allowNewlineParagraph` to false. There were still plenty of other errors... which are other false-positives of the check.

romani commented:
this Check code is very hard to read, lets refactor it a bit before this PR to proceeed.

please look at https://github.com/checkstyle/checkstyle/pull/15709 and try to finish it.
We need to split newline handling to make logic a bit more clear.

Zopsss commented:
build is still failing, there are still lots of errors

romani commented:
before we proceed on this, all other PRs and issues should be merged.
https://github.com/checkstyle/checkstyle/issues/15713 must be done before, as it will make inputs more realistic so we will make update more clearly.

romani commented:
@Zopsss , please resolve conflicts in this PR to let me start review.
you can create new Inputs to avoid conflicts with pending PR for Block-tag.

Zopsss commented:
> we need to use for checkstyle-checks.xml
> `<property name="allowNewlineParagraph" value="false"/>`

reverted this as we fixed this in one of my old PR.

Zopsss commented:
local build was still failing, it was mostly giving errors for:

```
<p> tag should be preceded with an empty line.
Redundant <p> tag.
```

It was also giving this error:

```
<p> tag should be placed immediately before the first word, with no space after.
```

for cases like:

```
<p> testing: </p>
   ↑ whitespace 
```

the above error is fine, it is violation of rule but other 2 errors are false-positive, they started appearing because we added support to check for open-close tag. 

Should we keep these errors or not? If yes, then it would require changing lots of files.

romani commented:
https://github.com/checkstyle/checkstyle/blob/37a777c3f69b162e88bb6fc88659fe4595248b29/config/checkstyle-checks.xml#L720

This means that allowNewlineParagraph= true

And it means that you have defect, as this property is to allow \n.

Zopsss commented:
> And it means that you have defect, as this property is to allow \n.

yes we allow it, so that check does not give violation on lines like:

https://github.com/checkstyle/checkstyle/blob/37a777c3f69b162e88bb6fc88659fe4595248b29/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L28-L31

the default property is what we need for our style. We set it to false for google_checks.xml because it does not follow such rules.

romani commented:
Google style use false, checkstyle style is true.
We need to improve logic for allowNewlineParagraph to do not violate our javadocs that are keep paragraph tag alone on line.

Zopsss commented on config/checkstyle-checks.xml:
```diff
@@ -46,6 +46,18 @@
     <property name="files"
               value=".*[\\/]src[\\/](test|it|xdocs-examples)[\\/].*(?&lt;!Support)\.java"/>
   </module>
+  <!--
+    Until we fully support opening-closing paragraph tag
+    https://github.com/checkstyle/checkstyle/issues/15685
+   -->
+  <module name="SuppressionSingleFilter">
+    <property name="checks" value="JavadocParagraph"/>
+    <property name="message" value="tag should be preceded with an empty line."/>
+  </module>
+  <module name="SuppressionSingleFilter">
+    <property name="checks" value="JavadocParagraph"/>
+    <property name="message" value="Redundant .* tag."/>
+  </module>
```
suppression until we fix the implementation of this check. May change the referenced issue in future if we open new issue for it

romani replied:
We need to fix almost all violations from Paragraph Check. https://github.com/checkstyle/checkstyle/blob/aeaf4b0838d814ce62ef23aa6b4d367ff5f83fb2/src/main/java/com/puppycrawl/tools/checkstyle/JavaAstVisitor.java#L84-L86 is good example. Please start a process of updating our sources, let's make PRs to be limited to ~20  files. All PRs will have "supplemental: " prefix.

romani replied:
If there is some strange cases, report to us and skip them for now, we will handle them at the end of as soon as you get confirmation on how to update javadoc.

romani replied:
we need fixes, please continue https://github.com/checkstyle/checkstyle/pull/15737 and all other fixes.
We eat food we produce before we release to it to users, we need to fix all of them.

romani replied:
Put this link in comment , you will update all in separate PR

Zopsss replied:
this will be addressed at https://github.com/checkstyle/checkstyle/issues/15762

Zopsss commented:
A PR has been sent for failing CI: https://checkstyle.semaphoreci.com/jobs/46bc3abc-9190-43d9-bb99-3a25a47a386b

https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/1037

Zopsss commented:
fixed some bug in the latest push: https://github.com/checkstyle/checkstyle/compare/643b58d3d4f9225d764a9a064a5aa5e5eac69503..c80a41ccedf6312677b6578b19091f815bb3d1e1#diff-19c14e011c79f51ad2e551f4941a1def0d5ee70c40d93fd944e07da166bad4bb

romani commented:
@Zopsss , please send PR to pgjdbc project to fix violations.

Zopsss commented:
Build is flooded with errors again :(

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/28127/workflows/60b27b2a-fb0d-4774-a64f-b37dea3364ca/jobs/698909

this time it is because of following type of violation:

```
<p> tag should not precede HTML block-tag '<ul>', <p> tag should be removed. [JavadocParagraph]
```

should I add suppression for this too?

Zopsss commented:
GitHub, generate report for configs in PR description

Zopsss commented:
GitHub, generate report for configs in PR description

Zopsss commented:
Fixed https://github.com/checkstyle/checkstyle/pull/15686#issuecomment-2391546197 in the latest push: https://github.com/checkstyle/checkstyle/compare/1042402e60a306559e7a2f263389b8ec7d91ddaa..de2acad15324993d03afdf25fac93db4dcbe9fc5

Zopsss commented:
GitHub, generate report for configs in PR description

romani commented on src/it/resources/com/google/checkstyle/test/chapter7javadoc/rule712paragraphs/InputIncorrectJavadocParagraph.java:
```diff
@@ -143,9 +139,12 @@ class InnerPrecedingPtag {
+    // 2 violations 6 lines above:
+    //  '<p> tag should be placed immediately before the first word'
+    //  '<p> tag should be preceded with an empty line.'
     public void foo() {}
```
this is actually false positives.
Google rule for P should be applicable only for P tags  that are not nested in other HTML tags.
in other words they are on root level of javadoc.

`<p> tag should be preceded with an empty line.` this is not html rule, html does not care about empty lines.
this rule is for humans who read javadoc as source, humans more easy to parse javadoc source in this case. I doubt empty line will do a difference in resulted rendered html from javadoc.

`'<p> tag should be placed immediately before the first word'` this is also not html related rule, it is just style of Google.

When first P is open, all nested in it, is HTML, and will have content to render well in browser, there could be any layout, we should not deal with it, it is very complicated and I doubt anybody care about this.

Lets keep our logic simple, if we detect P tag, and it is nested in other tag, we skip it.

Zopsss replied:
should we check if it precedes a block-tag or we should skip that too?

romani replied:
if P nested, no violations at all, no any violations.

romani replied:
please document this, users should know abou this.

romani reviewed (changes requested):
item:

romani commented:
CI failure:
```
[INFO] --- jacoco-maven-plugin:0.8.12:check (default-check) @ checkstyle ---
[INFO] Loading execution data file C:\projects\checkstyle\target\jacoco.exec
[INFO] Analyzed bundle 'checkstyle' with 837 classes
[WARNING] Rule violated for class com.puppycrawl.tools.checkstyle
  .checks.javadoc.JavadocParagraphCheck: branches covered ratio is 0.99, but expected minimum is 1.00
```

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:
```diff
@@ -0,0 +1,88 @@
+    * Some Summary.
+    *
+    * <p>
+    *   Some Javadoc.
```
this might be  false-negative.
Indentation spaces should trigger a violation.

this is no violation code:
```
    * <p>
    * Some Javadoc.
```

Zopsss replied:
This code was previously suggested by you:

https://github.com/checkstyle/checkstyle/blob/619265de69dabb63264cddaa387fc524c656a0f6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L213-L215

In this input file allowNewlineParagraph is set to true, based on the above code this example should not give violation if allowNewlineParagraph is set to true.

New lines are allowed, so I think this is correct code. `some javadoc` does not need to be right after `<p>`

Zopsss replied:
>* <p>
>    *   Some Javadoc.

If we give violation on this code and use `<p> should be placed immediately before text, without space` message then it would confuse users. Correct me if I'm wrong.

romani replied:
Lets not be blocked by this,  we might need to move this discussion on a side to separate issue.

romani replied:
please create issue

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag.java:
```diff
@@ -0,0 +1,88 @@
+    * </p>
+    *
+    * <p>
+    *     <h1>testing....</h1>
```
should be violation

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag2.java:
```diff
@@ -0,0 +1,94 @@
+    * </p>
+    *
+    * <p>
+    *     <h1>testing....</h1>
```
should be violation.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphIncorrectOpenClosedTag3.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+    * Some Summary.
+    *
+    * <p><!-- comment is required --> extra space in paragraph after comment is violation</p>
+    */
+    // violation 2 lines above '<p> tag should be placed immediately before the first word'
```
please update documentation that this Check ignores HTML comments.

Zopsss replied:
done https://github.com/checkstyle/checkstyle/blob/5e28cccd6ecf9074c5bca15e71ab4b258bd77539/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L44-L46

romani reviewed (changes requested):
please update doc to describe new behavior.

items:

Zopsss commented:
GitHub, generate report for configs in PR description

Zopsss commented:
GitHub, generate report for configs in PR description

romani commented:
Github, generate report for configs in PR description

romani commented:
GitHub, generate report for JavadocParagraph/all-examples-in-one

Zopsss commented:
I tried using the latest commit in validation.sh for pgjdbc

https://checkstyle.semaphoreci.com/jobs/4b4e6720-796f-4bb0-ad18-69345b97de77#L689

but it requires java 17 or 21, current version was 11: https://checkstyle.semaphoreci.com/jobs/4b4e6720-796f-4bb0-ad18-69345b97de77#L718

I looked into validation.sh to update the java version but couldn't find it.

romani commented:
GitHub, generate report for JavadocParagraph/all-examples-in-one

romani commented:
@Zopsss , please send PR to pgjdbc  for their latest master
and I am ok to just disable pgjdbc until this PR is merged, We will find way to use jdk17 in semaphore a bit later.

romani commented:
@Zopsss , please rebase over latest master, I upgraded pgjdbc to SHA that you have in this PR.

Zopsss commented:
PR sent at pgjdbc https://github.com/pgjdbc/pgjdbc/pull/3397

Zopsss commented:
@romani finally, all CIs are passing. I have fixed incorrect paragraphs at pgjdbc. Please do the final review.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocparagraph/InputJavadocParagraphCheck1.java:
```diff
@@ -31,4 +32,39 @@ class Check {
+     * Some summary.
+     *
+     *<p>
+     *  <ul>
```
This should be violation, why it is not ?
Even we ignore \n , there are bunch of spaces before UL.
And should be no P before UL, as it is blocktag

Zopsss replied:
> Even we ignore \n , there are bunch of spaces before UL.

we decided not to check this at https://github.com/checkstyle/checkstyle/pull/15686#discussion_r1788518551, read my comment above it for explanation

> And should be no P before UL, as it is blocktag

fixed this, thanks for pointing this out

romani replied:
please create issue and put "ok , until #xxxxxxxxx" to let readers know that we detected it but postponed this false-negative

romani replied:
Please put new issue in several case of this PR to make it clear it is postponed

Zopsss replied:
this will be addressed at https://github.com/checkstyle/checkstyle/issues/15762

Zopsss replied:
I have added `// ok until 15762` comments

romani reviewed (changes requested):
Items

romani commented:
GitHub, generate report for JavadocParagraph/all-examples-in-one

romani commented:
GitHub, generate report for JavadocParagraph/all-examples-in-one

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:
```diff
@@ -44,6 +44,9 @@
  * nested paragraph tags are allowed to do that.</li>
  * </ul>
+ * <p>
+ * <b>ATTENTION:</b> This Check ignores HTML comments.
```
we need to add something about "if P nested, no violations at all, no any violations."
see https://github.com/checkstyle/checkstyle/pull/15686/files#r1787687046

Zopsss replied:
done

https://github.com/checkstyle/checkstyle/blob/cb7cac3b2e167e9fc260dc64576c87ab45257cfd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L48-L53

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:
```diff
@@ -174,7 +177,8 @@ public void visitJavadocToken(DetailNode ast) {
         else if (ast.getType() == JavadocTokenTypes.HTML_ELEMENT
-                && JavadocUtil.getFirstChild(ast).getType() == JavadocTokenTypes.P_TAG_START) {
+                && (JavadocUtil.getFirstChild(ast).getType() == JavadocTokenTypes.P_TAG_START
+                || JavadocUtil.getFirstChild(ast).getType() == JavadocTokenTypes.PARAGRAPH)) {
```
add 4 spaced before `||` to put it visually under `(` to show priority of execution

Zopsss replied:
done

https://github.com/checkstyle/checkstyle/blob/cb7cac3b2e167e9fc260dc64576c87ab45257cfd/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java#L184-L188

romani reviewed (changes requested):
items:

Zopsss commented:
not sure why ci is failing

https://app.circleci.com/pipelines/github/checkstyle/checkstyle/28232/workflows/df53342b-c53d-431f-9c77-888fd26fff20/jobs/703338

local build was passing.

Zopsss commented:
Github, generate site

romani commented:
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/50771010/job/410wglmmfwmaa7nr#L254

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.12.1:site (default-site) on project checkstyle: Error parsing 'C:\projects\checkstyle\src\xdocs\checks\javadoc\javadocparagraph.xml': Error validating the model: Error:
[ERROR]   Public ID: null
[ERROR]   System ID: null
[ERROR]   Line number: 34
[ERROR]   Column number: 15
[ERROR]   Message: cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://maven.apache.org/XDOC/2.0":ul}'. One of 
```

romani commented on src/xdocs/checks/javadoc/javadocparagraph.xml.template:
```diff
@@ -29,6 +29,14 @@
         </ul>
+        <p>
+          <b>ATTENTION:</b>
+          <ul>
```
Hmm, this is not allowed tag ... Strange.
Can we workaround it somehow?
https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/50771010/job/410wglmmfwmaa7nr#L254

romani replied:
You can reproduce on local by `./.ci/validation.cmd site_without_verify`

Or `mvn site`

romani reviewed (changes requested):
Items

romani commented:
Github, generate site

Zopsss commented:
GitHub, generate report for JavadocParagraph/all-examples-in-one

romani reviewed (approved):
Ok to merge

rdiachenko commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocParagraphCheck.java:
```diff
@@ -198,25 +203,49 @@ private void checkEmptyLine(DetailNode newline) {
+     * @return true, if the paragraph tag is nested.
+     */
+    private static boolean isNestedParagraph(DetailNode tag) {
+        boolean isNested = false;
```
Please rename to `nested`, it's a variable name, shouldn't be a verb.

Zopsss replied:
done

rdiachenko reviewed (approved):
Minor note, ok to merge after it is fixed.

Zopsss commented:
Check no closed issue reference failing CI is addressed at https://github.com/checkstyle/checkstyle/pull/15765

romani reviewed (approved):
I see fix for above item.
